smithy-trait-codegen/src/main/java/software/amazon/smithy/traitcodegen/generators/ToNodeGenerator.java

```diff
@@ -142,22 +125,75 @@ public Void structureShape(StructureShape shape) {
                 writer.writeWithNoFormatting(".sourceLocation(getSourceLocation())");
             }
             for (MemberShape mem : shape.members()) {
-                if (TraitCodegenUtils.isNullableMember(mem)) {
-                    writer.write(".withOptionalMember($S, get$U().map(m -> $C))",
-                            mem.getMemberName(),
-                            symbolProvider.toMemberName(mem),
-                            (Runnable) () -> mem.accept(new ToNodeMapperVisitor("m")));
+                Shape memTarget = model.expectShape(mem.getTarget());
+                boolean isMemCollection = memTarget.isMapShape() || memTarget.isListShape();
+                boolean isNullable = TraitCodegenUtils.isNullableMember(mem);
+                String memberName = mem.getMemberName();
+                String getterName = symbolProvider.toMemberName(mem);
+                if (isMemCollection) {
+                    generateCollectionMember(mem, memTarget, memberName, getterName, isNullable);
                 } else {
-                    writer.write(".withMember($S, $C)",
-                            mem.getMemberName(),
-                            (Runnable) () -> mem.accept(new ToNodeMapperVisitor(symbolProvider.toMemberName(mem))));
+                    generateSimpleMember(mem, memberName, getterName, isNullable);
                 }
             }
             writer.writeWithNoFormatting(".build();");
             writer.dedent();
             return null;
         }
 
+        private void generateCollectionMember(
+                MemberShape mem,
```

Comment:
Let's use the full spelling here.
```suggestion
                MemberShape member,
```